src/ssh/scp_client.cpp

```diff
@@ -90,37 +100,50 @@ void mp::SCPClient::pull_file(const std::string& source_path, const std::string&
     SCPUPtr scp{make_scp_session(*ssh_session, SSH_SCP_READ, source_path.c_str())};
     SSH::throw_on_error(ssh_scp_init, scp);
 
-    auto r = ssh_scp_pull_request(scp.get());
-
-    if (r != SSH_SCP_REQUEST_NEWFILE)
-        throw std::runtime_error("Error receiving information for file");
-
-    auto size = ssh_scp_request_get_size(scp.get());
-    std::string filename{ssh_scp_request_get_filename(scp.get())};
-
-    auto total{0u};
-    const auto len{65536u};
-    std::vector<char> data;
-    data.reserve(len);
-
-    QFile destination(QString::fromStdString(destination_path));
-    if (!destination.open(QIODevice::WriteOnly))
-        throw std::runtime_error("Error opening file for writing: " + destination.errorString().toStdString());
+    int r;
 
-    SSH::throw_on_error(ssh_scp_accept_request, scp);
-
-    do
+    while ((r = ssh_scp_pull_request(scp.get())) != SSH_SCP_REQUEST_EOF)
     {
-        r = ssh_scp_read(scp.get(), data.data(), len);
-
-        if (r == 0)
-            break;
-
-        if (destination.write(data.data(), r) == -1)
-            throw std::runtime_error("Error writing to file: " + destination.errorString().toStdString());
-
-        total += r;
-    } while (total < size);
+        if (r == SSH_ERROR || r == SSH_SCP_REQUEST_WARNING)
+            throw std::runtime_error("Error receiving information for file: " + std::string(ssh_get_error(scp.get())));
+
+        auto size = ssh_scp_request_get_size(scp.get());
+        std::string filename{ssh_scp_request_get_filename(scp.get())};
+
+        auto total{0u};
+        const auto len{65536u};
+        std::vector<char> data;
+        data.reserve(len);
+
+        auto full_destination_path{destination_path};
+        if (full_destination_path.empty())
+        {
+            full_destination_path = filename;
+        }
+        else if (QFileInfo(QString::fromStdString(destination_path)).isDir())
+        {
+            full_destination_path.append("/" + filename);
+        }
```

Comment:
Since it's the same logic as in push_file, this should be its own function in the anonymous namespace.

Reply:
It's not quite the same logic in that in `pull_file`, it will use the filename retrieved from the scp server, but in `push_file`, it will the filename provided in the source path.  So then, we would still have to pass in the filename we want to use.  But yeah, probably some way to genericise this some...